File: 19-Remove-Nth-Node-From-End-of-List.py
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        curr = dummy

        for _ in range(n):
            head = head.next

        while head:
            head, curr = head.next, curr.next

        curr.next = curr.next.next
        return dummy.next

File: 19-Remove-Nth-Node-From-End-of-List.js
/**
 * Definition for singly-linked list.
 */
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val);
    this.next = (next === undefined ? null : next);
}

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
    let dummy = new ListNode(0),
        fast = dummy,
        slow = dummy;

    dummy.next = head;

    for (let i = 0; i <= n; i++)
        fast = fast.next;

    while (fast !== null) {
        fast = fast.next;
        slow = slow.next;
    }

    slow.next = slow.next.next;
    return dummy.next;
};
